drivers/clock_control/clock_control_rpi_pico.c

```diff
@@ -0,0 +1,885 @@
+		[RPI_PICO_CLKID_CLK_PERI] = {
+			.source = 0,
+			.aux_source = CLOCK_AUX_SOURCE(clk_peri),
+			.source_rate = SRC_CLOCK_FREQ(clk_peri),
+			.rate = CLOCK_FREQ(clk_peri),
+		},
+		[RPI_PICO_CLKID_CLK_USB] = {
+			.source = 0,
+			.aux_source = CLOCK_AUX_SOURCE(clk_usb),
+			.source_rate = SRC_CLOCK_FREQ(clk_usb),
+			.rate = CLOCK_FREQ(clk_usb),
+		},
+		[RPI_PICO_CLKID_CLK_ADC] = {
+			.source = 0,
+			.aux_source = CLOCK_AUX_SOURCE(clk_adc),
+			.source_rate = SRC_CLOCK_FREQ(clk_adc),
+			.rate = CLOCK_FREQ(clk_adc),
+		},
+		[RPI_PICO_CLKID_CLK_RTC] = {
+			.source = 0,
+			.aux_source = CLOCK_AUX_SOURCE(clk_rtc),
+			.source_rate = SRC_CLOCK_FREQ(clk_rtc),
+			.rate = CLOCK_FREQ(clk_rtc),
+		},
+	},
+	.plls_data = {
+		[RPI_PICO_PLL_SYS] = {
+			.ref_div = DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, pll_sys), clock_div),
+			.fb_div = DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, pll_sys), fb_div),
+			.post_div1 = DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, pll_sys), post_div1),
+			.post_div2 = DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, pll_sys), post_div2),
+		},
+		[RPI_PICO_PLL_USB] = {
+			.ref_div = DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, pll_usb), clock_div),
+			.fb_div = DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, pll_usb), fb_div),
+			.post_div1 = DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, pll_usb), post_div1),
+			.post_div2 = DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, pll_usb), post_div2),
+		},
+	},
+	.rosc_data = {
+		.phase = (COND_CODE_1(DT_NODE_HAS_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, rosc),
+						       phase_flip),
+				      (ROSC_PHASE_FLIP_BITS), (0x0)) |
+			  COND_CODE_1(DT_NODE_HAS_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, rosc),
+						       phase),
+				      ((DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, rosc), phase) &
+					ROSC_PHASE_SHIFT_BITS) | ROSC_PHASE_ENABLE_BITS), (0x0))),
+		.div = DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, rosc), clock_div),
+		.range = DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, rosc), range),
+		.code = (STAGE_DS(0) | STAGE_DS(1) | STAGE_DS(2) | STAGE_DS(3) |
+			 STAGE_DS(4) | STAGE_DS(5) | STAGE_DS(6) | STAGE_DS(7)),
+	},
+	.gpin_data = {
+		[RPI_PICO_GPIN_0] = {
+			COND_CODE_1(DT_NODE_HAS_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, gpin0),
+						 clock_frequency),
+				    (.frequency = DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, gpin0),
+							  clock_frequency),),
+				    (.frequency = 0,))
+		},
+		[RPI_PICO_GPIN_1] = {
+			COND_CODE_1(DT_NODE_HAS_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, gpin1),
+						     clock_frequency),
+				    (.frequency = DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, gpin1),
+							  clock_frequency),),
+				    (.frequency = 0,))
+		},
+	},
+};
+
+static struct clock_control_rpi_pico_data clock_control_rpi_pico_data = {
+	.rosc_freq = CLOCK_FREQ(rosc),
+	.rosc_ph_freq = CLOCK_FREQ(rosc_ph),
+};
+
+DEVICE_DT_INST_DEFINE(0, &clock_control_rpi_pico_init, NULL, &clock_control_rpi_pico_data,
+		      &clock_control_rpi_pico_config, PRE_KERNEL_1,
+		      CONFIG_CLOCK_CONTROL_INIT_PRIORITY, &clock_control_rpi_pico_api);
```

Comment:
Should the init priority really be configurable? Maybe this should just be a hard coded zero

Reply:
The priority is constrained strongly. But not we cannot be configured at all. We can configure it relative to other drivers' priority. And the default value will not cause trouble in usual case.
So, I leave as it is.